### PR TITLE
hds: prevent timer reset on every update

### PR DESCRIPTION
**Pull request by sschepens**
If we push frequent updates but with the same interval, we may never get responses back because the timer gets reset on every update.
This change makes a reset only if the interval has changed or a disconnect ocurred.

### CODE REVIEW

**Comment by sschepens** [comment]
@markatou @htuch could you take a look at this please?

**Comment by sschepens** [comment]
@htuch i was wondering how to test this in an integration test. Do you have any ideas?
I was thinking about about making an update loop, but then verify that reports have been received, but this seems somehow tricky.

**Comment by htuch** [comment]
@sschepens I think some variant of https://github.com/envoyproxy/envoy/blob/90a7a3eb8f477350a7175cc0d98487bbed1c4188/test/integration/hds_integration_test.cc#L341 with multiple updates and a sufficiently long timeout to cover them should work. Bonus points for using `Test/SimulatedTimeSystem` (not sure how easy it is to use inside integration tests yet).

**Comment by sschepens** [comment]
@htuch we have 2 things that could be tested, one is the update of the interval, if I change the interval I should receive a new response in the new interval. The other thing is the bug of the timer being reset everytime an update is received.
Not sure if we want to test both, testing an update on the interval is probably trivial, but testing the timer reset is somehow tricky, i was thinking of making a loop of update+responseCheck with a wait time lower than the interval, if the timer is reset everytime, this would generate a neverending loop.

**Comment by htuch** [comment]
@sschepens 

> i was thinking of making a loop of update+responseCheck with a wait time lower than the interval, if the timer is reset everytime, this would generate a neverending loop.

Maybe just iterate some arbitrary number of times, e.g. 10, and see if we observe 10 responses?

**Comment by sschepens** [comment]
@htuch 
> Maybe just iterate some arbitrary number of times, e.g. 10, and see if we observe 10 responses?

We would need to sleep enough in each iteration to allow for responses to arrive, would that be OK?

File: test/integration/hds_integration_test.cc [hunk 1]
```diff
@@ -687,5 +687,45 @@ TEST_P(HdsIntegrationTest, TestUpdateMessage) {
...
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  healthcheckEndpoints();
+
+  // an update should be received after interval
+  ASSERT_TRUE(
+      hds_stream_->waitForGrpcMessage(*dispatcher_, response_, std::chrono::milliseconds(100)));
+
+  // New HealthCheckSpecifier message
+  server_health_check_specifier_.mutable_interval()->set_nanos(300000000); // 0.3 seconds
```
Comment (htuch) [inline test/integration/hds_integration_test.cc]: Maybe make the time skew 250ms instead of 100ms between expect/timeout, that should hopefully not flake.

**Comment by htuch** [comment]
Interval update test looks good (although I wish we could use simulated time, will need to check with @jmarantz when he is back on this). I'd take as guidance that we don't want these tests to take more than a few seconds each as the upper bound on any real-time delays you'll add.

**Comment by jmarantz** [comment]
+1 for using simulated time, which is definitely done in at least one integration test:

https://github.com/envoyproxy/envoy/blob/50c2357aa0d7a943e404150f1c578cdb59326aa4/test/integration/integration_admin_test.cc#L454

I'm not 100% sure whether that will work in this case, because if the timing is done in grpc itself it won't be using the Envoy time system abstraction.

It's easy enough to tell: just switch to a SimulatedTimeSystem per the pattern above, and then run the test. It will run very fast, and probably fail 99% of the time, or work 100% of the time.

**Comment by sschepens** [comment]
> It's easy enough to tell: just switch to a SimulatedTimeSystem per the pattern above, and then run the test. It will run very fast, and probably fail 99% of the time, or work 100% of the time.

@jmarantz it would seem that `Event::SimulatedTimeSystem` conflicts with `BaseIntegrationTest` they both provide a `timeSystem()` getter, which is actually being used in `hds_integration_test`.
what would be the way to work around this?

**Comment by jmarantz** [comment]
See the example I pointed to. Just inherit from SimulatedTimeSystem in your test class.

TimeSystem is the base class for SimulatedTimeSystem.

**Comment by sschepens** [comment]
> Maybe make the time skew 250ms instead of 100ms between expect/timeout, that should hopefully not flake.

@htuch not excatly sure what you meant here, but I increased the timeouts to avoid flakes

**Comment by jmarantz** [comment]
@sschepens ping me on slack if you have more questions about how sim-time works with integration tests. The the admin integration test I described above inherits indirectly from BaseIntegrationTest but it just inherits from SimulatedTimeSystem first, which establishes what time-system to use for the duration of the test.

**Comment by htuch** [comment]
@sschepens Looks good, but could you chat with @jmarantz and figure out if we could use simulated time here without too much extra work? Thanks.

**Comment by jmarantz** [comment]
SimulatedTimeSystem works with this test, but actually (and surprisingly to me) makes it 10x slower. I want to go investigate why that is and follow up. In the meantime I suggest going forward with this as is, using real time and I can address later.

**Comment by htuch** [comment]
@jmarantz ack, thanks for looking into this folks.

**Pull request merged**